File: app/src/main/java/com/github/bjoernpetersen/q/ui/fragments/SongFragment.java
package com.github.bjoernpetersen.q.ui.fragments;

import android.content.Context;
import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.support.v7.widget.DividerItemDecoration;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.RecyclerView.ItemDecoration;
import android.support.v7.widget.RecyclerView.LayoutManager;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import com.github.bjoernpetersen.jmusicbot.client.model.Song;
import com.github.bjoernpetersen.q.R;
import java.util.ArrayList;
import java.util.List;

/**
 * A fragment representing a list of Items.
 * <p />
 * Activities containing this fragment MUST implement the {@link OnListFragmentInteractionListener}
 * interface.
 */
public class SongFragment extends Fragment {

  private static final String ARG_SONG_LIST = "song-list";
  private ArrayList<Song> songs;
  private OnListFragmentInteractionListener mListener;

  /**
   * Mandatory empty constructor for the fragment manager to instantiate the
   * fragment (e.g. upon screen orientation changes).
   */
  public SongFragment() {
  }

  @SuppressWarnings("unused")
  public static SongFragment newInstance(List<Song> songs) {
    SongFragment fragment = new SongFragment();
    Bundle args = new Bundle();
    args.putParcelableArrayList(ARG_SONG_LIST, new ArrayList<>(songs));
    fragment.setArguments(args);
    return fragment;
  }

  @Override
  public void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);

    if (getArguments() != null) {
      songs = getArguments().getParcelableArrayList(ARG_SONG_LIST);
    }

    if (songs == null) {
      songs = new ArrayList<>();
    }
  }

  @Override
  public View onCreateView(LayoutInflater inflater, ViewGroup container,
      Bundle savedInstanceState) {
    View view = inflater.inflate(R.layout.fragment_song_list, container, false);

    // Set the adapter
    if (view instanceof RecyclerView) {
      Context context = view.getContext();
      RecyclerView recyclerView = (RecyclerView) view;
      recyclerView.setAdapter(new SongRecyclerViewAdapter(songs, mListener));
      DividerItemDecoration decoration = new DividerItemDecoration(
          recyclerView.getContext(),
          DividerItemDecoration.VERTICAL
      );
      recyclerView.addItemDecoration(decoration);
    }
    return view;
  }


  @Override
  public void onAttach(Context context) {
    super.onAttach(context);
    if (context instanceof OnListFragmentInteractionListener) {
      mListener = (OnListFragmentInteractionListener) context;
    } else {
      throw new RuntimeException(context.toString()
          + " must implement OnListFragmentInteractionListener");
    }
  }

  @Override
  public void onDetach() {
    super.onDetach();
    mListener = null;
  }

  /**
   * This interface must be implemented by activities that contain this
   * fragment to allow an interaction in this fragment to be communicated
   * to the activity and potentially other fragments contained in that
   * activity.
   * <p/>
   * See the Android Training lesson <a href=
   * "http://developer.android.com/training/basics/fragments/communicating.html"
   * >Communicating with Other Fragments</a> for more information.
   */
  public interface OnListFragmentInteractionListener {

    void onAdd(Song song);

    boolean showAdd(Song song);
  }
}

File: app/src/main/java/com/github/bjoernpetersen/q/ui/fragments/SongFragment.kt
package com.github.bjoernpetersen.q.ui.fragments

import android.content.Context
import android.os.Bundle
import android.support.annotation.IdRes
import android.support.annotation.MenuRes
import android.support.v4.app.Fragment
import android.support.v7.widget.DividerItemDecoration
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import com.github.bjoernpetersen.jmusicbot.client.model.Song
import com.github.bjoernpetersen.q.R
import com.github.bjoernpetersen.q.ui.fragments.SongFragment.SongFragmentInteractionListener

/**
 * A fragment representing a list of Items.
 *
 *
 * Activities containing this fragment MUST implement the [SongFragmentInteractionListener]
 * interface.
 */
class SongFragment : Fragment() {

  private var songs: ArrayList<Song>? = null
  private var songContext: Int? = null
  private var mListener: SongFragmentInteractionListener? = null

  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)
    val args = arguments ?: throw IllegalStateException("Arguments are missing")
    songs = args.getParcelableArrayList(ARG_SONG_LIST) ?: ArrayList()
    songContext = if (args.containsKey(ARG_SONG_CONTEXT)) args.getInt(ARG_SONG_CONTEXT) else null
  }

  override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
      savedInstanceState: Bundle?): View? {
    val view = inflater.inflate(R.layout.fragment_song_list, container, false)

    val songs = this.songs ?: throw IllegalStateException()

    // Set the adapter
    if (view is RecyclerView) {
      view.adapter = SongRecyclerViewAdapter(songs, mListener, songContext)
      val decoration = DividerItemDecoration(view.context, DividerItemDecoration.VERTICAL)
      view.addItemDecoration(decoration)
    }
    return view
  }

  override fun onAttach(context: Context) {
    super.onAttach(context)
    if (context is SongFragmentInteractionListener) mListener = context
    else throw RuntimeException(
        context.toString() + " must implement SongFragmentInteractionListener")
  }

  override fun onDetach() {
    songs = null
    mListener = null
    super.onDetach()
  }

  /**
   * This interface must be implemented by activities that contain this
   * fragment to allow an interaction in this fragment to be communicated
   * to the activity and potentially other fragments contained in that
   * activity.
   *
   *
   * See the Android Training lesson [Communicating with Other Fragments](http://developer.android.com/training/basics/fragments/communicating.html) for more information.
   */
  interface SongFragmentInteractionListener {

    fun onClick(song: Song, enable: EnableCallback)
    fun isEnabled(song: Song): Boolean
    fun isEnabled(song: Song, @IdRes menuItemId: Int): Boolean
    fun onContextMenu(song: Song, menuItem: MenuItem, enable: (Boolean) -> Unit): Boolean
    fun onSongListTouch() = Unit
  }

  companion object {
    @JvmStatic
    private val ARG_SONG_LIST = SongFragment::class.java.name + ".songList"
    @JvmStatic
    private val ARG_SONG_CONTEXT = SongFragment::class.java.name + ".songContext"

    @JvmStatic
    fun newInstance(songs: List<Song>, @MenuRes songContext: Int? = null): SongFragment {
      val fragment = SongFragment()
      val args = Bundle()
      args.putParcelableArrayList(ARG_SONG_LIST, ArrayList(songs))
      if (songContext != null)
        args.putInt(ARG_SONG_CONTEXT, songContext)
      fragment.arguments = args
      return fragment
    }
  }
}
